### ISSUE TITLE
Export single-item cypher assumes that the related nodes do not exist, and this not always ideal

### ISSUE THREAD

**Issue by Joshfindit** [neo4j-contrib/neo4j-apoc-procedures#247]
Granted; I know this is a little tricky conceptually, but the idea will be to create all nodes first, then run the rel import.

First hiccup: exporting just the relationships is not possible as export requires a list of nodes, but this is minor (a simple post-processing change to `MERGE` in the export should be enough to work around this)

The real issue is related nodes with different labels...

Here's the query:

```
MATCH (a:Artefact)-[r]->(n)
WHERE a.uuid = "afac671b-4a9d-4245-869c-74f085f13e57" // Looking for a specific node and it's outgoing relationships
WITH collect(distinct a) as a, collect(r) as r // Required as apoc.export.cypherData needs a list, not just a single item
CALL apoc.export.cypher.data(a, r, '/tmp/cypherexports/single.cypher', {batchSize:1000})
YIELD nodes, relationships, properties, time
RETURN nodes, relationships, properties, time
```

The result:

```
begin
CREATE (:`Artefact` {`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57", `description`:"Description 1", `name`:"Artefact 1"});
commit
begin
CREATE CONSTRAINT ON (node:`Artefact`) ASSERT node.`uuid` IS UNIQUE;
commit
schema await
begin
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`Artefact`{`uuid`:"1d0a2d1d-53cc-481c-b794-afcc6d9c7323"}) CREATE (n1)-[:`REFERS_TO`]->(n2);
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`Artefact`{`uuid`:"bfe18c38-14d4-4a6f-97a1-e62b4021ad2a"}) CREATE (n1)-[:`REFERS_TO`]->(n2);
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`Artefact`{`uuid`:"43c73548-d2f8-4e32-b201-a4c94cc9bd88"}) CREATE (n1)-[:`REFERS_TO`]->(n2);
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`Artefact`{`uuid`:"78a8df58-c667-4e2c-a238-0e8b0823ed1b"}) CREATE (n1)-[:`REFERS_TO`]->(n2);
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`Artefact`{`uuid`:"8e74c8ab-7057-4632-8e34-f5fc09b47142"}) CREATE (n1)-[:`REFERS_TO`]->(n2);
MATCH (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"}), (n2:`UNIQUE IMPORT LABEL`{`UNIQUE IMPORT ID`:61}) CREATE (n1)-[:`PERMISSION_TO`]->(n2);
commit
```

The issue here is the last line.
If we wipe the database, create all nodes, and then run this Cypher, the last relationship will never be made properly (and worse, may connect to something unrelated)
Last line *should* be ```MATCH (n1:`Artefact`{`uuid`:"f322b2ec-89a9-4502-a243-879ffa53b157"}), (n2:`Identity`{`uuid`:"257a1954-a25c-4376-bd22-b48eef00753c"}) CREATE (n1)-[:`PERMISSION_TO`]->(n2);```

**Comment by Joshfindit** [comment]
(Another datapoint: also tried `call apoc.export.cypher.query("MATCH (a:Artefact)-[r]->(n) RETURN r", "/tmp/cypherexports/queryexport.cypher", {batchSize:10});`, and got `CREATE` statements for all related `a` and `n` nodes)

**Comment by jexp** [comment]
What does your graph look like?

I recently fixed a bug in this area which apoc version do you use?

**Comment by Joshfindit** [comment]
I'm growing it later, but right now it's just
```
(Artefact)-[:REFERS_TO]->(Artefact)
(Artefact)-[:PERMISSION_TO]->(Identity)
(Identity)-[:SESSION]->(WebSession)
```
Version: `apoc-3.1.0.3-all.jar`

**Comment by Joshfindit** [comment]
Would it help at this stage if I provided any additional information?

**Comment by Joshfindit** [comment]
There are a couple different issues that come up while doing 'single node' exports.
If I were to export an entire graph in to individual files containing `(node)` and an optional match on `(n)-[r]->(m)` (return n, r), it makes sense that importing all files would recreate the entire graph, but it's not working that way.

Here's a breakdown:
1. CREATE instead of MERGE (happens on at least `cypher.data` and `cypher.query`)
2. Adding 'extra' nodes in to the file (Create on `(m)` nodes) (Also on `.data` and `.query`)
3. (the biggest issue) that bug on the last relationship

Of note: It seems to be a choice between 2 and 3, where exports either have extra nodes or relationships defined by internal node id.

**Comment by inserpio** [comment]
@jexp is it ok to close this issue to work on these 3 points:

> 1. CREATE instead of MERGE (happens on at least cypher.data and cypher.query)
> 2. Adding 'extra' nodes in to the file (Create on (m) nodes) (Also on .data and .query)
> 3. (the biggest issue) that bug on the last relationship

**Comment by AngeloBusato** [comment]
With current apoc version 3.1.3.6 with this dataset:
```
CREATE (n1:`Artefact`{`uuid`:"afac671b-4a9d-4245-869c-74f085f13e57"})
CREATE (n2:`Artefact`{`uuid`:"1d0a2d1d-53cc-481c-b794-afcc6d9c7323"})
CREATE (n3:`Artefact`{`uuid`:"bfe18c38-14d4-4a6f-97a1-e62b4021ad2a"})
CREATE (n4:`Artefact`{`uuid`:"43c73548-d2f8-4e32-b201-a4c94cc9bd88"})
CREATE (n5:`Artefact`{`uuid`:"78a8df58-c667-4e2c-a238-0e8b0823ed1b"})
CREATE (n6:`Artefact`{`uuid`:"8e74c8ab-7057-4632-8e34-f5fc09b47142"})
CREATE (n1)-[:`REFERS_TO`]->(n2)
CREATE (n1)-[:`REFERS_TO`]->(n3)
CREATE (n1)-[:`REFERS_TO`]->(n4)
CREATE (n1)-[:`REFERS_TO`]->(n5)
CREATE (n1)-[:`REFERS_TO`]->(n6)
CREATE (n7:`Identity`{`uuid`:"257a1954-a25c-4376-bd22-b48eef00753c"})
CREATE (n1)-[:`PERMISSION_TO`]->(n7)
```
for 
```
MATCH (a:Artefact)-[r]->(n)
WHERE a.uuid = "afac671b-4a9d-4245-869c-74f085f13e57" // Looking for a specific node and it's outgoing relationships
WITH collect(distinct a) as a, collect(r) as r // Required as apoc.export.cypherData needs a list, not just a single item
CALL apoc.export.cypher.data(a, r, '/tmp/cypherexports/single.cypher', {batchSize:1000})
YIELD nodes, relationships, properties, time
RETURN nodes, relationships, properties, time
```
we have:
[export_cypher_data.txt](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/905746/export_cypher_data.txt)

while with
```
CALL apoc.export.cypher.query("MATCH (a:Artefact)-[r]->(n) WHERE a.uuid = 'afac671b-4a9d-4245-869c-74f085f13e57' return collect(distinct a) as a, collect(r) as r", 'single_query.cypher', {batchSize:1000})
YIELD nodes, relationships, properties, time
RETURN nodes, relationships, properties, time
```
we have:
[export_cypher_query.txt](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/905745/export_cypher_query.txt)

If you want the data about the other nodes you should pass them as param, like this:
```
MATCH (a:Artefact)-[r]->(n)
WHERE a.uuid = "afac671b-4a9d-4245-869c-74f085f13e57" 
WITH collect(distinct a)+collect(distinct n) as a, collect(r) as r 
CALL apoc.export.cypher.data(a, r, 'single_with_n.txt', {batchSize:10000})
YIELD nodes, relationships, properties, time
RETURN nodes, relationships, properties, time
 ```
[single_with_n.txt](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/905817/single_with_n.txt)

So for us all exports looks to be ok.

Merge vs Create is still open. 
@jexp should we introduced a new parameter for it? Otherwise we would close the issue.

**Comment by Joshfindit** [comment]
Assuming the tests work out the way the above code does, that resolves 3 (thank you).

It doesn't seem to address 2, however. The idea being that if the export assumes that related nodes already exist, there should only be 1 `CREATE`(/`MERGE`) in any given export file.

An example use-case is:

- 3 graphs that agree non data structure, but hold some different nodes. As each node can be globally unique, the graphs assume that a single UUID refers to a specific node, no matter which graph it's in.
- Referring to a single node, it is only present on graph 1, but relates to nodes that are present on all graphs
- Export the single node from graph 1
- Import it in to graph 2, connecting to the existing nodes since they already exist

A related use-case is copying single nodes between different versions of the same graph.

**Comment by jexp** [comment]
Let's create a new issue for adding a config option to export to use MERGE (both for nodes and rels) + `ON CREATE SET x += {key:"value",...}` (except the id-property)

**Comment by AngeloBusato** [comment]
@jexp for us this is fixed and merge case will be managed in the issue #380, right? we put it on review

**Comment by conker84** [comment]
Fixed by #380 feel free to reopen in case you need it
